File: src/aoc/2015/day-1/main.py
from pathlib import Path

def get_basement_direction_position(lines: list[str]):
  floor_number = 0
  for line in lines:
    for index, ltr in enumerate(line):
      floor_number += get_direction_char_value(ltr)
      if floor_number == -1:
        return index + 1
  raise ValueError("Unable to Find Basement Direction Position")

def get_direction_char_value(char: str):
  if char == "(":
    return 1
  elif char == ")":
    return -1
  return 0

def get_file_lines(file_path: str):
  file = Path(__file__).with_name(file_path)
  return file.read_text().splitlines()

def get_floor_number_from_directions(lines: str):
  floor_number = 0
  for line in lines:
    for ltr in line:
      floor_number += get_direction_char_value(ltr)
  return floor_number

INPUT_FILE_PATH = "./input.txt"

if __name__ == "__main__":
  directions = get_file_lines(INPUT_FILE_PATH)
  floor_number = get_floor_number_from_directions(directions)
  position = get_basement_direction_position(directions)
  print(f"Floor Number: {floor_number} | Basement Direction Position: {position}")
  

File: src/aoc/2015/day-1/main.cpp
#include <string>
#include <vector>
#include <fstream>
#include <iostream>

const std::string FILE_PATH = "./input.txt";

std::vector<std::string> get_file_lines(std::string file_path) {
    std::vector<std::string> lines;
    std::ifstream file(file_path);

    if (file.is_open()) {
        std::string line;
        while (getline(file, line)) {
            lines.push_back(line);
        }
        file.close();
    } else {
        printf("Unable to open file: %s\n", file_path.c_str());
    }

    return lines;
}

int main() {
    std::vector<std::string> lines = get_file_lines(FILE_PATH);
    std::string line = lines[0]; // file has only one line

    int floor = 0;
    int basement_index = -1;

    for (int index = 0; index < line.length(); index++) {
        char direction = line[index];

        if (direction == '(') floor++;
        else if (direction == ')') floor--;

        if (floor == -1 && basement_index == -1) {
            basement_index = index + 1;
        }
    }

    printf("Floor Number: %d\n", floor);
    printf("Basement Direction Position: %d\n", basement_index);

    return 0;
}